src/common/buffer.cc

```diff
@@ -1968,6 +1968,59 @@ int buffer::list::write_fd(int fd) const
   return 0;
 }
 
+int buffer::list::write_fd(int fd, uint64_t offset) const
+{
+  // use writev!
+  iovec iov[IOV_MAX];
+  int iovlen = 0;
+  ssize_t bytes = 0;
+
+  std::list<ptr>::const_iterator p = _buffers.begin();
+  while (p != _buffers.end()) {
+    if (p->length() > 0) {
+      iov[iovlen].iov_base = (void *)p->c_str();
+      iov[iovlen].iov_len = p->length();
+      bytes += p->length();
+      iovlen++;
+    }
+    ++p;
+
+    if (iovlen == IOV_MAX-1 ||
+	p == _buffers.end()) {
+      iovec *start = iov;
+      int num = iovlen;
+      ssize_t wrote;
+    retry:
```

Comment:
this impl isn't very elegant enough to me. Maybe we can refer to the alike impl(https://github.com/ceph/ceph/blob/master/src/msg/simple/Pipe.cc#L2359). It will be more readable